src/Icicle/Source/Checker/Constraint.hs

```diff
@@ -419,11 +399,27 @@ generateP scrutTy (pat, alt)
         require (annotOfExp alt) $ CEquals datS t
 
         (alt',sub) <- generateX alt
+        let altTy'  = annResult (annotOfExp alt')
+        altTy      <- TypeVar <$> fresh
+        altTp      <- TypeVar <$> fresh
+
+        -- Require alternative types to have the same temporality if they
+        -- do have temporalities. Otherwise defaults to TemporalityPure.
+        require (annotOfExp alt) $ CExtractTemporality altTp altTy altTy'
 
-        e' <- lift $ lift State.get
+        -- Require the alternative types without temporality to be the same.
+        require (annotOfExp alt) $ CEquals resTy altTy
+
+        -- Require return temporality to be compatible with alternative temporalities.
+        resTp' <- TypeVar <$> fresh
+        require (annotOfExp alt) $ CTemporalityJoin resTm resTp' altTp
+
+        e'    <- lift $ lift State.get
         lift $ lift $ State.put (e' { stateEnvironment = stateEnvironment e })
+        rest' <- generateP ann scrutTy resTy resTp' rest
+
+        return (((pat,alt'), sub):rest')
```

Comment:
hmm right, okay. I thought it would work because I thought everything that generates a temporality returned them already split up, eg as ``Temporality t u``.
ok, this is good then - maybe I'll have to fix the rest of the typechecker to perform this extraction too